### PR TITLE
client-go: add request and response size metrics

### PR DESCRIPTION
**Pull request by aojea**
/kind feature

#### What this PR does / why we need it:

```release-note
client-go: add new histogram metric to record the size of the requests and responses.
```

### CODE REVIEW

**Review by logicalhan — COMMENTED** [review]
Just one concern but this otherwise lgtm

File: staging/src/k8s.io/component-base/metrics/prometheus/restclient/metrics.go
Comment (aojea) [reply staging/src/k8s.io/component-base/metrics/prometheus/restclient/metrics.go]: oh, wait, this is a new interface, we can pass the host directly

File: staging/src/k8s.io/component-base/metrics/prometheus/restclient/metrics.go
Comment (dgrisonnet) [inline staging/src/k8s.io/component-base/metrics/prometheus/restclient/metrics.go:46]:
```suggestion
			Help: "Request size in bytes. Broken down by verb and host.",
```

**Review by logicalhan — APPROVED** [review]
/lgtm
/approve

File: staging/src/k8s.io/component-base/metrics/prometheus/restclient/metrics.go
Comment (aojea) [inline staging/src/k8s.io/component-base/metrics/prometheus/restclient/metrics.go:48]: the buckets are ok?, maybe for client-go we should have lower values?

File: staging/src/k8s.io/client-go/tools/metrics/metrics.go
Comment (logicalhan) [reply staging/src/k8s.io/client-go/tools/metrics/metrics.go:47]: It's fine to pass the URL and use the host property.

File: staging/src/k8s.io/component-base/metrics/prometheus/restclient/metrics.go
Comment (logicalhan) [reply staging/src/k8s.io/component-base/metrics/prometheus/restclient/metrics.go:48]: This is fine to me.

File: staging/src/k8s.io/component-base/metrics/prometheus/restclient/metrics.go
Comment (aojea) [inline staging/src/k8s.io/component-base/metrics/prometheus/restclient/metrics.go:45]: hmm, I think response is a better name

Comment (aojea) [reply staging/src/k8s.io/component-base/metrics/prometheus/restclient/metrics.go:45]: yeah, we should measure both , there are cases that the problems can be related to POST or GET with huge payloads,

File: staging/src/k8s.io/component-base/metrics/prometheus/restclient/metrics.go
Comment (aojea) [inline staging/src/k8s.io/component-base/metrics/prometheus/restclient/metrics.go:48]: @sttts @wojtek-t are these values ok ?

I vaguely remember that some place limits sizes ....

File: staging/src/k8s.io/client-go/tools/metrics/metrics.go
Comment (aojea) [inline staging/src/k8s.io/client-go/tools/metrics/metrics.go:45]: host :)

File: staging/src/k8s.io/component-base/metrics/prometheus/restclient/metrics.go
Comment (aojea) [reply staging/src/k8s.io/component-base/metrics/prometheus/restclient/metrics.go]: those are very high values, I changed for other more commons values

**Comment by aojea** [comment]
/hold cancel

image

image


request and response size graphs, @logicalhan please LGTM if you are good with these changes

**Comment by aojea** [comment]
hmm the unit test failures are legit :cry:

File: staging/src/k8s.io/component-base/metrics/prometheus/restclient/metrics.go
Comment (logicalhan) [inline staging/src/k8s.io/component-base/metrics/prometheus/restclient/metrics.go:48]: you forgot the stability level.

Comment (logicalhan) [reply staging/src/k8s.io/component-base/metrics/prometheus/restclient/metrics.go:48]: it seems it's missing on the older metric above, can you add it there too?

**Comment by aojea** [comment]
fixed, it was the empty r.c.base parameter on tests

File: staging/src/k8s.io/component-base/metrics/prometheus/restclient/metrics.go
Comment (aojea) [reply staging/src/k8s.io/component-base/metrics/prometheus/restclient/metrics.go:48]: ups

Comment (aojea) [reply staging/src/k8s.io/component-base/metrics/prometheus/restclient/metrics.go:50]: yeah, I've added it on both

**Comment by aojea** [comment]
> Test_Run_OneVolumeDetachFailNodeWithReadWriteOnce



/test pull-kubernetes-unit

File: staging/src/k8s.io/component-base/metrics/prometheus/restclient/metrics.go
Comment (dgrisonnet) [inline staging/src/k8s.io/component-base/metrics/prometheus/restclient/metrics.go:61]: Should we perhaps sync the buckets of `apiserver_response_sizes` based on these new ones as a follow-up? https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go#L125-L134

Comment (aojea) [reply staging/src/k8s.io/component-base/metrics/prometheus/restclient/metrics.go:61]: I think that this ones are more "realistic" , I'm +1 on that

**Review by logicalhan — COMMENTED** [review]
/lgtm

**Comment by aojea** [comment]
> /approve

I see @liggitt active, let's see if we can put this in his queue , wojtek told me he is a bit busy these days

File: staging/src/k8s.io/client-go/rest/request.go
Comment (liggitt) [inline staging/src/k8s.io/client-go/rest/request.go:904]: is ContentLength always populated? is there ever a case we should ignore (-1 or 0 with non-nil body?)

File: staging/src/k8s.io/client-go/rest/request.go
Comment (liggitt) [inline staging/src/k8s.io/client-go/rest/request.go:967]: does this do the right thing if result is an error result? (result.err != nil)

Comment (aojea) [reply staging/src/k8s.io/client-go/rest/request.go:967]: I had the same doubt for latency, and asking people with more experience they told me that latency should take into account errors too.
However, for size, I can't see why do we want to store a zero on a failed request, we didn't have a result so I don't think we should measure a length of nothing

File: staging/src/k8s.io/client-go/rest/request.go
Comment (aojea) [reply staging/src/k8s.io/client-go/rest/request.go:904]: I honestly can't say, I was trying to skim the golang code but found it confusing, I will add an if to avoid the -1 and 0 corner cases to be safe

File: staging/src/k8s.io/client-go/rest/request.go
Comment (aojea) [reply staging/src/k8s.io/client-go/rest/request.go:967]: result.body is nil on failure, hence I don't think we should measure it,

File: staging/src/k8s.io/client-go/rest/request.go
Comment (aojea) [reply staging/src/k8s.io/client-go/rest/request.go]: the condition is weird, but is based on 
https://cs.opensource.google/go/go/+/refs/tags/go1.17.7:src/net/http/request.go;l=919-927;drc=e6dda19888180c5159460486d30c0412e4980748

**Comment by fedebongio** [comment]
/triage accepted
Thank you to all the reviewers!

File: staging/src/k8s.io/client-go/rest/request.go
Comment (aojea) [reply staging/src/k8s.io/client-go/rest/request.go:988]: the metrics is associated to the response, and there is no response if err != nil, that is different from result.err != nil

File: staging/src/k8s.io/client-go/rest/request.go
Comment (aojea) [reply staging/src/k8s.io/client-go/rest/request.go:971]: I confirmed there is no null value in prometheus , so this is correct, only measure if it exists

File: staging/src/k8s.io/client-go/rest/request.go
Comment (aojea) [reply staging/src/k8s.io/client-go/rest/request.go]: > is ContentLength always populated? is there ever a case we should ignore (-1 or 0 with non-nil body?)

ignored on that case

File: staging/src/k8s.io/client-go/rest/request.go
Comment (liggitt) [reply staging/src/k8s.io/client-go/rest/request.go:971]: it still seems like this should be guarded on something like `if result.err == nil` of `if result.body != nil`

File: staging/src/k8s.io/client-go/rest/request.go
Comment (liggitt) [reply staging/src/k8s.io/client-go/rest/request.go:988]: should this be guarded on `if result.body != nil`?

Comment (liggitt) [reply staging/src/k8s.io/client-go/rest/request.go:988]: I think if `result.err != nil and len(result.body) == 0`, we should not record

File: staging/src/k8s.io/client-go/rest/request.go
Comment (aojea) [reply staging/src/k8s.io/client-go/rest/request.go]: ack

File: staging/src/k8s.io/client-go/rest/request.go
Comment (aojea) [reply staging/src/k8s.io/client-go/rest/request.go]: ack

**Comment by aojea** [comment]
/retest
E2eNode Suite: [sig-node] Summary API [NodeConformance] when querying /stats/summary should report resource usage through the stats api expand_less	
k8s.io/kubernetes/vendor/k8s.io/apiextensions-apiserver/test/integration: TestCRValidationOnCRDUpdate expand_less

**Comment by aojea** [comment]
E2eNode Suite: [sig-node] Summary API [NodeConformance] when querying /stats/summary should report resource usage through the stats api
/test pull-kubernetes-node-e2e-containerd

**Comment by aojea** [comment]
Kubernetes e2e suite: [sig-network] Proxy version v1 A set of valid responses are returned for both pod and service ProxyWithPath [Conformance] expand_less	
/test pull-kubernetes-conformance-kind-ga-only-parallel

**Pull request merged**
